src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsWriter.java

```diff
@@ -79,6 +84,9 @@ public KnnFieldVectorsWriter<?> addField(final FieldInfo fieldInfo) throws IOExc
     @Override
     public void flush(int maxDoc, final Sorter.DocMap sortMap) throws IOException {
         flatVectorsWriter.flush(maxDoc, sortMap);
+
+        initQuantizationStateWriterIfNecessary();
```

Comment:
So this will still open and write the header without checking if quantization is necessary
- you can check it inside the method `or`
- you do it after line 252 as the first step